src/storage/src/boundary/tcp_boundary.rs

```diff
@@ -326,41 +323,31 @@ pub mod server {
         sender: UnboundedSender<R>,
         /// A function to convert the input data into something the sender can transport.
         convert: F,
-        /// A token that's dropped once `client_token` is dropped.
-        token: Option<Rc<dyn Any>>,
         /// A weak reference to a token that e.g. the network layer can drop.
-        client_token: Weak<()>,
+        client_token: Weak<dyn Any + Send + Sync>,
     }
 
     impl<R, F> UnboundedEventPusher<R, F> {
         /// Construct a new pusher. It'll retain a weak reference to `client_token`.
         fn new(
             sender: UnboundedSender<R>,
-            token: Rc<dyn Any>,
-            client_token: &Arc<()>,
+            client_token: &Arc<dyn Any + Send + Sync>,
             convert: F,
         ) -> Self {
             Self {
                 sender,
                 client_token: Arc::downgrade(client_token),
                 convert,
-                token: Some(token),
             }
         }
     }
 
     impl<T, D, R, F: Fn(EventCore<T, D>) -> R> EventPusherCore<T, D> for UnboundedEventPusher<R, F> {
         fn push(&mut self, event: EventCore<T, D>) {
-            if self.client_token.upgrade().is_none() {
-                self.token.take();
-            }
-            if self.token.is_some() {
-                match self.sender.send((self.convert)(event)) {
-                    Err(_) => {
-                        self.token.take();
-                    }
-                    _ => {}
-                }
+            if self.client_token.upgrade().is_some() {
+                // Loss of the receiving end means we need to do nothing,
+                // so we ignore errors
+                let _ = self.sender.send((self.convert)(event));
```

Comment:
I think we may want to send things anyhow. Maybe a future bug to fix, but if we have dropped this, we will shut the source, and may want to communicate that onward (e.g. to allow downstream readers to shut down).

Reply:
I can add a todo, its unclear to me how we can force a send if the receiver is dead

Reply:
I'm not sure we need to do anything other than not suppress the send. In the case that the source shuts down, e.g. in response to the `client_token` being dropped, we may still want to send the resulting progress information to `sender`. If it weren't for the fact that COMPUTE pre-drops its imported sources, it would otherwise wait for that information that would otherwise not arrive.

Maybe best just to leave it for now, as we hope to be GCing it up soon.